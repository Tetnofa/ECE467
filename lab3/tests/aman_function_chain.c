int func1(int x) {
    return (x + 1);
}

int func2(int x, bool y) {
    if (y) {
        // return (func1(x) * 2);
    }
    return func1(x);
}

bool func3(int x, int y) {
    return (func1(x) > func1(y));
}

void main() {
    int result;
    result = func2(func1(5), func3(3, 2));
    result = func2(func1(5), func1(4));
    writeInt(result);
} 