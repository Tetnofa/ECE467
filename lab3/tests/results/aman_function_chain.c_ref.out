Program [useIO=0](1,0)
	Function[isProto=0]	Int(1,0)
		Identifier[name:func1](1,4)
		Parameter		Int(1,10)
			Identifier[name:x](1,14)
		Scope(1,17)
			ReturnStmt(2,4)
				Int Expression(2,12)
					Binary Expression[opcode:+](2,12)
						Int Expression(2,12)
							Reference(2,12)
								Identifier[name:x](2,12)
						Int Expression(2,16)
							IntConstant[val=1](2,16)
	Function[isProto=0]	Int(5,0)
		Identifier[name:func2](5,4)
		Parameter		Int(5,10)
			Identifier[name:x](5,14)
		Parameter		Bool(5,17)
			Identifier[name:y](5,22)
		Scope(5,25)
			IfStmt[hasElse=0](6,4)
				Int Expression(6,8)
					Reference(6,8)
						Identifier[name:y](6,8)
				Scope(6,11)
			ReturnStmt(9,4)
				CallExpr(9,11)
					Identifier[name:func1](9,11)
					Argument(9,17)
						Int Expression(9,17)
							Reference(9,17)
								Identifier[name:x](9,17)
	Function[isProto=0]	Bool(12,0)
		Identifier[name:func3](12,5)
		Parameter		Int(12,11)
			Identifier[name:x](12,15)
		Parameter		Int(12,18)
			Identifier[name:y](12,22)
		Scope(12,25)
			ReturnStmt(13,4)
				Bool Expression(13,12)
					Binary Expression[opcode:>](13,12)
						CallExpr(13,12)
							Identifier[name:func1](13,12)
							Argument(13,18)
								Int Expression(13,18)
									Reference(13,18)
										Identifier[name:x](13,18)
						CallExpr(13,23)
							Identifier[name:func1](13,23)
							Argument(13,29)
								Int Expression(13,29)
									Reference(13,29)
										Identifier[name:y](13,29)
	Function[isProto=0]	Void(16,0)
		Identifier[name:main](16,5)
		Scope(16,12)
			Scalar Declaration			Int(17,4)
				Identifier[name:result](17,8)
			Assignment(18,4)
				Reference(18,4)
					Identifier[name:result](18,4)
				CallExpr(18,13)
					Identifier[name:func2](18,13)
					Argument(18,19)
						CallExpr(18,19)
							Identifier[name:func1](18,19)
							Argument(18,25)
								Int Expression(18,25)
									IntConstant[val=5](18,25)
					Argument(18,29)
						CallExpr(18,29)
							Identifier[name:func3](18,29)
							Argument(18,35)
								Int Expression(18,35)
									IntConstant[val=3](18,35)
							Argument(18,38)
								Int Expression(18,38)
									IntConstant[val=2](18,38)
			Assignment(19,4)
				Reference(19,4)
					Identifier[name:result](19,4)
				CallExpr(19,13)
					Identifier[name:func2](19,13)
					Argument(19,19)
						CallExpr(19,19)
							Identifier[name:func1](19,19)
							Argument(19,25)
								Int Expression(19,25)
									IntConstant[val=5](19,25)
					Argument(19,29)
						CallExpr(19,29)
							Identifier[name:func1](19,29)
							Argument(19,35)
								Int Expression(19,35)
									IntConstant[val=4](19,35)
			Expression Statement(20,4)
				CallExpr(20,4)
					Identifier[name:writeInt](20,4)
					Argument(20,13)
						Int Expression(20,13)
							Reference(20,13)
								Identifier[name:result](20,13)
sema: 19:13 : no matching definition for func2
sema: 19:4 : type mismatch 
sema: 20:4 : use of undefined identifier writeInt
